projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php

```diff
@@ -0,0 +1,196 @@
+ */
+function wpcom_add_subscribe_modal_to_frontend() {
+	if ( should_enable_subscriber_modal() && ! is_admin() ) {
+		$posts = get_posts(
+			array(
+				'post_type'   => 'wp_template_part',
+				'post_status' => 'publish',
+				'numberposts' => -1,
+			)
+		);
+
+		$subscribe_template = reset(
+			array_filter(
+				$posts,
+				function ( $post ) {
+					return $post->post_name === get_subscribe_template_slug();
+				}
+			)
+		);
+
+		$blocks = parse_blocks( $subscribe_template->post_content );
+
+		/*
+		 * Jetpack requires escaping html for the render_block() call below.
+		 * I'm using wp_kses() and specifying the allowed tags.
+		 * But this is fragile: users can edit the Subscribe Modal
+		 * template and add tags that are not allowed here.
+		 */
+		$allowed_html = array(
+			'div'    => array(
+				'class' => array(),
+				'style' => array(),
+			),
+			'h2'     => array(
+				'class' => array(),
+				'style' => array(),
+			),
+			'p'      => array(
+				'type'  => array(),
+				'class' => array(),
+				'style' => array(),
+				'id'    => array(),
+			),
+			'form'   => array(
+				'action'                 => array(),
+				'method'                 => array(),
+				'accept-charset'         => array(),
+				'data-blog'              => array(),
+				'data-post_access_level' => array(),
+				'id'                     => array(),
+			),
+			'label'  => array(
+				'id'    => array(),
+				'for'   => array(),
+				'class' => array(),
+			),
+			'input'  => array(
+				'id'          => array(),
+				'type'        => array(),
+				'name'        => array(),
+				'value'       => array(),
+				'class'       => array(),
+				'style'       => array(),
+				'placeholder' => array(),
+				'required'    => array(),
+			),
+			'button' => array(
+				'type'  => array(),
+				'class' => array(),
+				'style' => array(),
+			),
+		);
+
+		?>
+			<div id="wpcom-subscribe-modal">
+				<div class="modal-content">
+					<?php
+					foreach ( $blocks as $block ) {
+						echo wp_kses( render_block( $block ), $allowed_html );
+					}
+					?>
+					<div id="close"><a href="#">Close</a></div>
+				</div>
+			</div>
+		<?php
+	}
+}
+add_action( 'wp_footer', 'wpcom_add_subscribe_modal_to_frontend' );
+
+/**
+ * Enqueues JS to load modal.
+ *
+ * @return void
+ */
+function wpcom_enqueue_subscribe_modal_assets() {
+	if ( should_enable_subscriber_modal() && ! is_admin() ) {
+		wp_enqueue_style( 'subscribe-modal-css', plugins_url( 'subscribe-modal.css', __FILE__ ), array(), JETPACK__VERSION );
+		wp_enqueue_script( 'subscribe-modal-js', plugins_url( 'subscribe-modal.js', __FILE__ ), array(), JETPACK__VERSION, true );
+	}
+}
+add_action( 'wp_enqueue_scripts', 'wpcom_enqueue_subscribe_modal_assets' );
+
+/**
+ * Returns true if we should enable Newsletter content.
+ * This is currently limited to lettre theme or newsletter sites.
+ * We could open it to all themes or site intents.
+ *
+ * @return bool
+ */
+function should_enable_subscriber_modal() {
```

Comment:
We can also limit to WP.com on first release?

Rest of the limits won't make sense for Jetpack sites, but for .com they could be reasonable start. And then release a bit wider next up?

Reply:
Maybe we could add a filter in here, so folks wanting to test can use the filter to toggle on the feature?

Reply:
@jeherve - interesting thought. Can you clarify or maybe link to an example of how that might work in jetpack for a situation like this? Not sure if we'd do this via, say, constant in wp-config or an actual WP filter or some other mechanism?

Reply:
Basically something like this:

```diff
diff --git a/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php b/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php
index 332abe2165..b77837781c 100644
--- a/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php
+++ b/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php
@@ -187,7 +187,16 @@ class Jetpack_Subscribe_Modal {
 		$is_modal_enabled   = get_option( 'sm_enabled' ) || false;
 		$is_fse_theme       = wp_is_block_theme();
 
-		return ( $is_lettre || $is_newsletter_site ) && $is_fse_theme && $is_modal_enabled;
+		$should_enable_subscriber_modal = ( $is_lettre || $is_newsletter_site ) && $is_fse_theme && $is_modal_enabled;
+
+		/**
+		 * Allows force-enabling or disabling the Subscriptions modal.
+		 *
+		 * @since $$next-version$$
+		 *
+		 * @param bool $should_enable_subscriber_modal Whether the Subscriptions modal should be enabled.
+		 */
+		return (bool) apply_filters( 'jetpack_subscriptions_modal_enabled', $should_enable_subscriber_modal );
 	}

```

You would then be able to force-enable the modal by adding the following in a functionality plugin:

```php
add_filter( 'jetpack_subscriptions_modal_enabled', '__return_true' );
```

Reply:
@jeherve - Thanks for the details. I've added this. I find myself wondering if we should also add a filter for users to add their own modal content, vs just using the new subscribe modal template part.  I'd do that later as a separate PR.

As a side note, a lot of my recent dev work has been on WordPress.com, and in that environment we think a lot less about leveraging these WP apis to make code behavior customizable by other devs. So this was a great general lesson to be thinking much more about that when working on Jetpack or other code that will be delivered via WordPress.org.

Reply:
I think it's fine if they can modify the modal via template editor and no need to modify via code.